### PR TITLE
proc/dwarf: Move LEB128 encode/decode functions to a separate package

### PR DESCRIPTION
**Pull request by alexandear**
This PR moves encode/decode LEB128 functions out from the `util` package to the `leb128` to improve maintainability. And because `util` is too general package name.

### CODE REVIEW

**Review by aarzilli — COMMENTED** [review]
I'm not sure this makes sense. If we could do away entirely with this package, sure, but just splitting off the LEB128 functions... what does it do in practice? Who invented the util = bad meme? What even is the justification for that?

**Comment by alexandear** [comment]
It's a common Go convention to avoid `utils` packages. We can read about package naming on the [official Go blog](https://go.dev/blog/package-names) and [Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#package-names) wiki article. They both said:

> Avoid meaningless package names like `util`, common, misc, api, types, and interfaces.

And have an explanation:

> Avoid meaningless package names. Packages named `util`, common, or misc provide clients with no sense of what the package contains. This makes it harder for clients to use the package and makes it harder for maintainers to keep the package focused. Over time, they accumulate dependencies that can make compilation significantly and unnecessarily slower, especially in large programs. And since such package names are generic, they are more likely to collide with other packages imported by client code, forcing clients to invent names to distinguish them.

Also, Dave Cheney has a [blog post](https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common) on that:
> A common cause of poor package names are `utility` packages. These are packages where helpers and utility code congeal. These packages contain an assortment of unrelated functions, as such their utility is hard to describe in terms of what the package provides. This often leads to a package’s name being derived from what the package contains—utilities.

That's why the PR was created.

**Comment by aarzilli** [comment]
Ok, but what are we going to do with ReadString? Make a package with just ReadString in it? Maybe we should just copy the contents of pkg/dwarf/util into pkg/dwarf.

**Comment by alexandear** [comment]
This PR doesn't touch other `util` functions and types: `ReadString`, `ReadUintRaw`, `WriteUint`, `ReadDwarfLengthVersion`, `ReadUnitVersions`, `UnknownFormat`, `MakeBuf`. These constructs and types can be moved in following pull requests. And can copy some of these functions to `pkg/dwarf`.

Actually, it's no problem with creating a package with only one function or variable. The standard library has such examples:  [color/palette](https://pkg.go.dev/image/color/palette@go1.19.4), [diff](https://pkg.go.dev/internal/diff@go1.19.4#Diff) (internal package).

**Comment by aarzilli** [comment]
> Actually, it's no problem with creating a package with only one function or variable

I just disagree with this.

**Review by derekparker — APPROVED** [review]
While I do agree with @aarzilli in that this is somewhat bikeshedding, I do find value in the improved organization and documentation for the package. I also appreciate you deprecating the existing functionality instead of outright removing it, as it gives folks who may rely on these packages externally a change to update their code. We will definitely include a note in the next release concerning this.

Overall I'm +1 on it, however given @aarzilli hesitancy I would like to wait for his +1 as well before we merge. I do believe if we merge this we should take things to the next step and organize the rest of the code as well, in my opinion I think it likely just belongs in the `pkg/dwarf` package.

**Comment by aarzilli** [comment]
IMO there's nothing wrong with util packages but if we really want to eliminate it then we should do it for real:
* the contents of buf.go can go in pkg/dwarf/godwarf which is the only user (as far as I can see), they also do not need to be exported
* the rest of the package could be moved to pkg/dwarf

I'd move the leb128 into pkg/dwarf as well, `dwarf.DecodeULEB128(...)` is fine enough as far as I am concerned, doesn't need to be in its own package.

**Comment by derekparker** [comment]
> IMO there's nothing wrong with util packages but if we really want to eliminate it then we should do it for real:
> 
> * the contents of buf.go can go in pkg/dwarf/godwarf which is the only user (as far as I can see), they also do not need to be exported
> * the rest of the package could be moved to pkg/dwarf
> 
> I'd move the leb128 into pkg/dwarf as well, `dwarf.DecodeULEB128(...)` is fine enough as far as I am concerned, doesn't need to be in its own package.

Agreed but I'm actually ok with leaving the leb128 package as it is, broken out on its own. It makes sense to group it together and lends itself nicely to being it's own package, with package documentation, etc.

As for the rest, I agree we should just make the switch. I still think though that it can be done in a follow-up PR. @alexandear would you be willing to take that on once this gets merged?

**Comment by alexandear** [comment]
@derekparker yes, I will do.

**Comment by derekparker** [comment]
@alexandear merged, feel free to send in the follow-up PR.

**Pull request merged**
